File: content/snippets/scala3/design_patterns/iterator.scala
// Iterator
trait CustomIterator[T]:
    def hasNext: Boolean
    def next(): T

trait CustomCollection[T]:
    def createIterator(): CustomIterator[T]

class ArrayIterator[T](items: Vector[T]) extends CustomIterator[T]:
    private var position = 0
    
    def hasNext: Boolean = position < items.length
    
    def next(): T =
        val result = items(position)
        position += 1
        result

class ArrayCollection[T](items: Vector[T]) extends CustomCollection[T]:
    def createIterator(): CustomIterator[T] = new ArrayIterator(items)

@main def main(): Unit =
    val collection = new ArrayCollection(Vector(10, 20, 30))
    val it = collection.createIterator()
    
    while (it.hasNext) {
        print(it.next() + " ")
    }
    println()

File: content/snippets/cpp/design_patterns/iterator.cpp
#include <iostream>
#include <vector>
#include <memory>

template <typename T>
class Iterator {
public:
    virtual ~Iterator() = default;
    virtual bool hasNext() = 0;
    virtual T next() = 0;
};

template <typename T>
class Collection {
public:
    virtual ~Collection() = default;
    virtual std::unique_ptr<Iterator<T>> createIterator() = 0;
};

template <typename T>
class ArrayIterator : public Iterator<T> {
private:
    std::vector<T>& items;
    int position = 0;
public:
    ArrayIterator(std::vector<T>& items) : items(items) {}
    
    bool hasNext() override {
        return position < items.size();
    }
    
    T next() override {
        return items[position++];
    }
};

template <typename T>
class ArrayCollection : public Collection<T> {
private:
    std::vector<T> items;
public:
    void add(T item) { items.push_back(item); }
    
    std::unique_ptr<Iterator<T>> createIterator() override {
        return std::make_unique<ArrayIterator<T>>(items);
    }
};

int main() {
    ArrayCollection<int> collection;
    collection.add(10);
    collection.add(20);
    collection.add(30);
    
    auto it = collection.createIterator();
    while (it->hasNext()) {
        std::cout << it->next() << " ";
    }
    std::cout << std::endl;
}
